compiler/luci/service/src/CircleShapeInferenceRule.test.cpp

```diff
@@ -48,6 +49,54 @@ bool shape_pass(loco::Graph *g)
   return loco::apply(&rules).to(g);
 }
 
+bool is_same_shape(luci::CircleNode *node, loco::TensorShape shape)
+{
+  if (node->shape_status() != luci::ShapeStatus::VALID)
+    return false;
+
+  if (node->rank() != shape.rank())
+    return false;
+
+  for (uint32_t i = 0; i < node->rank(); ++i)
+  {
+    if (node->dim(i).known() != shape.dim(i).known())
+      return false;
+
+    if (node->dim(i).value() != shape.dim(i).value())
+      return false;
+  }
+
+  return true;
+}
+
+// NOTE This function imitates CircleShapeInferencePass but little bit different.
+//      In CircleShapeInferencePass, DeadNodeQueryService is used to get alive nodes,
+//      which are not detected with postorder_traversal.
+//      However, it is not considered in this function because this is just for testing
+//      inference rule itself, not for inference pass.
+bool circle_shape_pass(loco::Graph *g)
+{
+  luci::sinf::Rule shape_infer_rule;
+  bool changed = false;
+
+  for (auto node : loco::postorder_traversal(loco::output_nodes(g)))
+  {
+    loco::TensorShape shape;
+    auto circle_node = loco::must_cast<luci::CircleNode *>(node);
+
+    if (shape_infer_rule.infer(circle_node, shape) && !is_same_shape(circle_node, shape))
+    {
+      circle_node->rank(shape.rank());
+      for (uint32_t i = 0; i < shape.rank(); ++i)
+        circle_node->dim(i) = shape.dim(i);
+      circle_node->shape_status(luci::ShapeStatus::VALID);
```

Comment:
(optional) maybe add this as a function...  maybe not in this PR...
Where... in `luci/IR/CircleNodeShape.h` or `luci/Service/CircleNodeShape.h`
I think `luci/IR/CircleNodeShape.h` seems better. what do you think?

Reply:
(optional)  `template <> class NodeMixin<NodeTrait::TensorShape>` in compiler/loco/include/loco/IR/NodeMixins.h,
we can add another shape setter like
```cxx
  void shape(std::initializer_list<uint32_t> dims)
  {
    rank(dims.size());

    uint32_t axis = 0;
    for (auto d : dims)
    {
      dim(axis++) = d;
    }
  }

  void shape(const loco::TensorShape &shape) or (const loco::TensorShape *shape)
  {
    ...
  }
```

Reply:
If there is no problem to add following function at `loco/IR/NodeMixins.h`, it would be good :)

```cpp
  void shape(const loco::TensorShape &shape)
  {
    ...
  }
```

Both introducing `luci/IR/CircleNodeShape.h` and adding following function at `luci/IR/CircleNodeDecl.h` looks good to me!
```cpp
struct CircleNode : public loco::Node,
                    public loco::NodeMixin<loco::NodeTrait::DataType>,
                    public loco::NodeMixin<loco::NodeTrait::TensorShape>
{
  const loco::TensorShape &shape(void) const { return _shape; }
  void shape(const loco::TensorShape &shape) { rank(shape.rank); ... }
...
}
```

Reply:
> and adding following function at luci/IR/CircleNodeDecl.h looks good to me!

FYI, there is no `_shape`. `public loco::NodeMixin<loco::NodeTrait::TensorShape>` provides the attributes of `class NodeMixin<NodeTrait::TensorShape>`